transfer/transfer.go

```diff
@@ -0,0 +1,177 @@
+}
+
+type Client interface {
+	// Transport returns the transport used by the transfer
+	Transport() transport.Transport
+	// PVCs returns the list of PVCs the transfer will migrate
+	PVCs() []*corev1.PersistentVolumeClaim
+	// IsCompleted returns whether the client is done
+	Status(ctx context.Context, c client.Client) (*Status, error)
+	// MarkForCleanup adds a key-value label to all the resources to be cleaned up
+	MarkForCleanup(ctx context.Context, c client.Client, key, value string) error
+}
+
+// PodOptions allow callers to pass custom configuration for the transfer pods
+type PodOptions struct {
+	// PodSecurityContext determines what GID the rsync process gets
+	// In case of shared storage SupplementalGroups is configured to get the gid
+	// In case of block storage FSGroup is configured to get the gid
+	PodSecurityContext corev1.PodSecurityContext
+	// ContainerSecurityContext determines what selinux labels, UID and drop capabilities
+	// are required for the containers in rsync transfer pod via SELinuxOptions, RunAsUser and
+	// Capabilities fields respectively
+	ContainerSecurityContext corev1.SecurityContext
+	// NodeName allows pods to be scheduled on a specific node. This is especially required in
+	// client pods where the PVC's are bound to a specific regions and the node where the pod runs on
+	// has to be in the same region as the PVC to be scheduled and bound.
+	NodeName string
+	// NodeSelector is a wider net for scheduling the pods on node than NodeName.
+	NodeSelector map[string]string
+	// Resources allows for configuring the resources consumed by the transfer pods. In general
+	// it is good to provision destination transfer pod with same or larger resources than the source
+	// so that the network is not congested.
+	Resources corev1.ResourceRequirements
+}
+
+type Status struct {
+	Running   *Running
+	Completed *Completed
+}
+
+type Running struct {
+	StartedAt *metav1.Time
+}
+
+type Completed struct {
+	Successful bool
+	Failure    bool
+	FinishedAt *metav1.Time
+}
+
+// IsPodHealthy is a utility function that can be used by various
+// implementations to check if the server pod deployed is healthy
+func IsPodHealthy(ctx context.Context, c client.Client, pod client.ObjectKey) (bool, error) {
+	p := &corev1.Pod{}
+
+	err := c.Get(context.Background(), pod, p)
+	if err != nil {
+		return false, err
+	}
+
+	return areContainersReady(p)
+}
+
+// IsPodCompleted is a utility function that can be used by various
+// implementations to check if the server pod deployed is completed.
+// if containerName is empty string then it will check for completion of
+// all the containers
+func IsPodCompleted(ctx context.Context, c client.Client, podKey client.ObjectKey, containerName string) (bool, error) {
+	pod := &corev1.Pod{}
+	err := c.Get(context.Background(), podKey, pod)
+	if err != nil {
+		return false, err
+	}
+
+	if len(pod.Status.ContainerStatuses) != 2 {
```

Comment:
The assumption was that one container is for transport and another for transfer, Although this might likely change if there is no transport involved in the future.